### PR TITLE
New evaluation and costing

### PR DESCRIPTION
**Pull request by aslesarenko**
> This new implementation of ErgoTree v2 is based on #637,  #697 
The code of this PR can be merged into v3.x/v4.0 branches, BUT NOT ACTIVATED. 
The actual activation should be done in v5.0 release as it is described in https://github.com/ScorexFoundation/sigmastate-interpreter/blob/develop/docs/aot-jit-switch.md.
That said, this PR MUST NOT introduce consensus breaking changes in the existing (AOT based) implementation of the interpreter.

### Motivation
ErgoTree is simple declarative intermediate representation for Ergo contracts.
It is designed to be compact in serialized form and directly executable, i.e. no additional transformation is necessary before it can be efficiently executed. 

This PR implements big-step recursive ErgoTreeEvaluator that works directly with ErgoTree.
Because of this the evaluator is very simple and follows [denotational semantics of ErgoTree](https://ergoplatform.org/docs/ErgoTree.pdf).
Or, the other way around, this implementation of ErgoTreeEvaluator is purely functional with immutable data structures and can be used as definition of ErgoTree's semantics.

### Implementation
ErgoTreeEvaluator takes ErgoTree directly as it is deserialized as part of a transaction.
No additional transformations is performed.
ErgoTree is interpreted directly and all intermediate data are stored in runtime types.
Runtime types are such types as Coll, SigmaProp, AvlTree, BigInt, etc.
It also use immutable Map to keep current DataEnv of computed ValDefs, as a result only the addition operation is used from the map, and the deletion is essentially done by the garbage collection.

### Performance
Since this interpreter directly works with SigmaDsl types (Coll, BigInt, SigmaProp etc), it turns out to be fast.
Since it also does JIT style costing instead of AOT style, it is 5-6x faster then existing implementation.

### Milestone
New evaluation of ErgoTree required changes in how isProven handled.
In particular isProven is not supported in ErgoTree that is passed to prove/verify methods.
This required changes in tests and also in the coverage report.
Unfortunately those changes may break the consensus, because isProven can be serialized and present in ErgoTree.
This prevents ErgoTreeEvaluator to be used in v3.x

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    (res, cost)
+  }
+
+  def error(msg: String) = sys.error(msg)
+
+  def msgCostLimitError(cost: Long, limit: Long) = s"Estimated execution cost $cost exceeds the limit $limit"
+
+}
+
+/** Incapsulate simple monotonic (add only) counter with reset. */
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:86]: Encapsulate

File: docs/notes.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
+## Costing
```
Comment (kushti) [inline docs/notes.md:1]: Better move it into Cost.md (already exists) or new-costing.md

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import sigmastate.lang.exceptions.CostLimitException
+import special.sigma.Context
+
+case class EvalSettings(isMeasureOperationTime: Boolean)
+
+class EvalContext(
+  val context: Context,
+  val constants: Seq[Constant[SType]])
+
+class ErgoTreeEvaluator(
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:18]: Please provide ScalaDoc

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 4]
```diff
@@ -34,6 +35,9 @@ trait Interpreter extends ScorexLogging {
   val IR: IRContext
   import IR._
 
+  val evalSettings: EvalSettings = ErgoTreeEvaluator.DefaultEvalSettings
+  val returnAOTCost: Option[Boolean] = None
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:39]: Please provide ScalaDoc

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 5]
```diff
@@ -139,34 +143,66 @@ trait Interpreter extends ScorexLogging {
     * @return          result of script reduction
     * @see `ReductionResult`
     */
-  def reduceToCrypto(context: CTX, env: ScriptEnv, exp: Value[SType]): Try[ReductionResult] = Try {
+  def reduceToCrypto(context: CTX, env: ScriptEnv, exp: Value[SType], treeComplexity: Int): Try[ReductionResult] = Try {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:146]: Please update ScalaDoc

File: docs/notes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply docs/notes.md:1]: moved to new-costing.md

Comment (aslesarenko) [reply docs/notes.md:1]: Done

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:125]: Done

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:52]: Done

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 1 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:125]: Done

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 3 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:52]: Done

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 4 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:39]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 5 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:146]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  * The runtime types are such types as [[special.collection.Coll]],
+  * [[special.sigma.SigmaProp]], [[special.sigma.AvlTree]], [[BigInt]], etc.
+  * It also use immutable Map to keep current [[DataEnv]] of computed [[ValDef]]s, as
+  * result only addition is used from the map, and deletion is essentially a garbage
+  * collection.
+  *
+  * ### Performance
+  * Since this interpreter directly works with SigmaDsl types (Coll, BigInt, SigmaProp
+  * etc), it turns out to be very fast. Since it also does JIT style costing instead of
+  * AOT style, it is 5-6x faster then existing implementation.
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:50]: then => than

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+/** Configuration parameters of the evaluation run. */
+case class EvalSettings(
+  /** Used by [[ErgoTreeEvaluator.profiler]] to measure operations timings. */
+  isMeasureOperationTime: Boolean,
+  /** Used by [[ErgoTreeEvaluator]] to conditionally perform debug mode operations. */
+  isDebug: Boolean = false)
+
+/** Used as a wrapper around Sigma [[Context]] to be used in [[ErgoTreeEvaluator]].
+  */
+class EvalContext(val context: Context)
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:21]: Why to use the wrapper, not Context directly?

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  * Since this interpreter directly works with SigmaDsl types (Coll, BigInt, SigmaProp
+  * etc), it turns out to be very fast. Since it also does JIT style costing instead of
+  * AOT style, it is 5-6x faster then existing implementation.
+  */
+class ErgoTreeEvaluator(
+  /** Represents data context for ErgoTree evaluation */
+  val evalContext: EvalContext,
+  /** Accumulates computation costs. */
+  val coster: CostAccumulator,
+  /** Preforms operations profiling and time measurments (if enabled in settings). */
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:57]: Preforms => Performs, measurments => measurements

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+{
+  /** */
+  def eval(env: DataEnv, exp: SValue): Any = {
+    exp.evalTo[Any](this, env)
+  }
+
+  def addCostOf(opName: String, opType: SFunc) = {
+    val cost = Value.costOf(opName, opType)
+    coster.add(cost)
+//    println(s"CostOf($opName, $opType) -> $cost") // comment before commit and push
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:70]: Please avoid commented out code, there are much better solutions for conditional logic.

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  /** Immutable data environment used to assign data values to graph nodes. */
+  type DataEnv = Map[Int, Any]
+
+  /** A profiler which is used by default if [[EvalSettings.isMeasureOperationTime]] is enabled. */
+  val DefaultProfiler = new Profiler
+
+  /** Default global [[EvalSettings]] instance. */
+  val DefaultEvalSettings = EvalSettings(isMeasureOperationTime = false)
+
+  /** Evaluate the given [[ErgoTree] in the given Ergo context using the given settings.
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:102]: Unclosed [ tag

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 7 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:21]: @kushti This is because EvalContext is wider than the blockchain Context. In particular current implementation doesn't support direct evaluation of ConstantPlaceholder nodes, which require access to ErgoTree segregated constants from the `eval` method. So the plan is to add those constants in the EvalContext. It has already been there, but I removed it recently. (https://github.com/ScorexFoundation/sigmastate-interpreter/pull/639/commits/dd7ef08ff0d27ab2718e2d6079d8e500b13fb8d7#diff-0c0204307ccbd881996df87f58c6078e)

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 6 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:50]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 8 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:57]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 9 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:70]: @kushti which one you keep in mind?

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 10 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:102]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 9 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala]: fixed

File: docs/notes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kushti) [reply docs/notes.md]: new-cosing.md actually, which is not a proper name

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 7 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:23]: It is totally unclear from the class definition, made comments then

File: docs/notes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply docs/notes.md]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 7 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:23]: Even comments are not a good solution I guess, you need somehow explicitly specify in the code (with comments) how EvalContext is wider.  From the definition it looks like an alias.

File: docs/new-cosing.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
+## Costing Rules
```
Comment (kushti) [inline docs/new-cosing.md:1]: Should be moved to a separate PR on top of this branch. Such unfinished documents usually stay unfinished for long time with no proper control.

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala [hunk 12]
```diff
@@ -118,17 +119,16 @@ class ErgoLikeContext(val lastBlockUtxoRoot: AvlTreeData,
       dataBoxes, boxesToSpend, newSpendingTransaction, selfIndex, extension, validationSettings, costLimit, initCost)
 
 
-  override def toSigmaContext(IR: Evaluation, isCost: Boolean, extensions: Map[Byte, AnyValue] = Map()): sigma.Context = {
-    implicit val IRForBox: Evaluation = IR
+  override def toSigmaContext(isCost: Boolean, extensions: Map[Byte, AnyValue] = Map()): sigma.Context = {
```
Comment (kushti) [inline sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:122]: If you touch old code, better add comments on the way

File: sigmastate/src/main/scala/sigmastate/eval/Profiling.scala [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+import sigmastate.serialization.OpCodes
+import sigmastate.serialization.OpCodes.OpCode
+import sigmastate.serialization.ValueSerializer.getSerializer
+import scalan.util.Extensions.ByteOps
+import scala.collection.mutable
+import debox.{Map => DMap}
+
+class OpStat(val node: SValue, val outerStart: Long, var innerTime: Long, val outerEnd: Long)
+
+class Profiler {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/eval/Profiling.scala:15]: ScalaDoc

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala [hunk 14]
```diff
@@ -28,4 +31,9 @@ class ErgoLikeInterpreter(implicit val IR: IRContext) extends Interpreter {
         }.orElse(d.default)
       case _ => super.substDeserialize(context, updateContext, node)
     }
+
+  def reduceToCrypto2(context: CTX, env: ScriptEnv, exp: SigmaPropValue): Try[ReductionResult] = Try {
```
Comment (kushti) [inline sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala:35]: ScalaDoc

File: sigmastate/src/main/scala/sigmastate/utils/SparseArrayContainer.scala [hunk 15]
```diff
@@ -1,7 +1,7 @@
 package sigmastate.utils
 
 import sigmastate.SType
-import sigmastate.Values.Value
+import sigmastate.Values.{Value, ValueCompanion}
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/utils/SparseArrayContainer.scala:4]: Why unused import added ???

File: sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala [hunk 16]
```diff
@@ -170,13 +171,15 @@ object CostTable {
     ("BinAnd", "(Boolean, Boolean) => Boolean", logicCost),
     ("BinOr", "(Boolean, Boolean) => Boolean", logicCost),
     ("BinXor", "(Boolean, Boolean) => Boolean", logicCost),
-    ("AND", "(Coll[Boolean]) => Boolean", logicCost),
+    ("OR", "(Coll[Boolean]) => Boolean", logicBaseCost),
     ("OR_per_item", "(Coll[Boolean]) => Boolean", logicCost),
+    ("AND", "(Coll[Boolean]) => Boolean", logicBaseCost),
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala:176]: Looks like consensus critical issue, cost reduced for an operation in 3.x

File: sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala [hunk 17]
```diff
@@ -84,6 +84,7 @@ object CostTable {
   val comparisonPerKbCost = 10
 
   val logicCost = 10
+  val logicBaseCost = 0  // TODO soft-fork: increase it to 20
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala:87]: Please add description what is this at all

File: sigmastate/src/main/scala/sigmastate/utxo/transformers.scala [hunk 18]
```diff
@@ -22,9 +26,14 @@ case class MapCollection[IV <: SType, OV <: SType](
                                                     mapper: Value[SFunc])
   extends Transformer[SCollection[IV], SCollection[OV]] {
   override def companion = MapCollection
-  implicit def tOV = mapper.asValue[OV].tpe
   override val tpe = SCollection[OV](mapper.tpe.tRange.asInstanceOf[OV])
   override val opType = SCollection.MapMethod.stype.asFunc
+  protected final override def eval(E: ErgoTreeEvaluator, env: DataEnv): Any = {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/utxo/transformers.scala:31]: ScalaDoc for this and all the other touched methods

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 19]
```diff
@@ -34,6 +35,19 @@ trait Interpreter extends ScorexLogging {
...
   import IR._
 
+  /** Evaluation settings used by [[ErgoTreeEvaluator]] */
+  def evalSettings: EvalSettings = ErgoTreeEvaluator.DefaultEvalSettings
+
+  /** Specifies which cost to use as result of evaluation.
+    * Used only in tests when evalSettings.isDebug == true.
+    * The meaning of values:
+    * - when Some(true) return AOT based cost
+    * - when Some(false) return AOT based cost
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:45]: I guess JIT here, not AOT?

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 20]
```diff
@@ -217,13 +271,91 @@ trait Interpreter extends ScorexLogging {
       val prop = propositionFromErgoTree(tree, context1)
 
       implicit val vs = context1.validationSettings
-      val (propTree, context2) = trySoftForkable[(BoolValue, CTX)](whenSoftFork = (TrueLeaf, context1)) {
+      val (propTree, context2) = trySoftForkable[(SigmaPropValue, CTX)](whenSoftFork = (TrueLeaf, context1)) {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:274]: Please add comments

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 21]
```diff
@@ -217,13 +271,91 @@ trait Interpreter extends ScorexLogging {
...
+        applyDeserializeContext(context1, prop)
+      }
+
+      // here we assume that when `propTree` is TrueProp then `reduceToCrypto` always succeeds
+      // and the rest of the verification is also trivial
+      val (cProp, cost) = reduceToCrypto(context2, env, propTree, (context2.initCost - context.initCost).toInt).getOrThrow
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:280]: Looks like consensus-related issue, isn't it ?

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 7 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:23]: EvalContext is removed

File: docs/new-cosing.md (hunk 11 above)
Comment (aslesarenko) [reply docs/new-cosing.md:1]: removed

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala (hunk 12 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala:122]: added ScalaDoc in the ancestor trait

File: sigmastate/src/main/scala/sigmastate/eval/Profiling.scala (hunk 13 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/Profiling.scala:15]: Done

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala (hunk 14 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala:35]: removed

File: sigmastate/src/main/scala/sigmastate/utils/SparseArrayContainer.scala (hunk 15 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utils/SparseArrayContainer.scala:4]: removed

File: sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala (hunk 17 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala:87]: added ScalaDoc

File: sigmastate/src/main/scala/sigmastate/utxo/transformers.scala (hunk 18 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utxo/transformers.scala:31]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 19 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:45]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 20 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:274]: added ScalaDoc for trySoftForkable

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 21 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:280]: Which issue you mean?

Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:281]: change in the cost calculated. then old Sigma-3.x powered node and updated one will diverge in block cost estimation, which could result in disagreement on block validation and network split as result.

File: sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala (hunk 17 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala:91]: Seems for 

"("AND", "(Coll[Boolean]) => Boolean", logicBaseCost)" breaking cost change was introduced

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 21 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:281]: no, all the historical blocks will be valid with respect to JIT. This will be ensured during [forking process](https://github.com/ergoplatform/eips/blob/5f9d14dabe38828f083f72b88bf828e2e78d7dd6/eip-0008.md#selection-of-hard-forking-block).

Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:281]: @aslesarenko I don't see any conditional logic around "context2.initCost - context.initCost"

File: sigmastate/src/main/scala/sigmastate/lang/Terms.scala [hunk 22]
```diff
@@ -113,6 +117,26 @@ object Terms {
...
+        val argV = args(0).evalTo[Any](E, env)
+        fV(argV)
+      }
+      else {
+        // TODO coverage
+        val f = func.evalTo[Seq[Any] => Any](E, env)
+        val argsV = args.map(a => a.evalTo[Any](E, env))
+        f(argsV)
+      }
+      // TODO JITC
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/lang/Terms.scala:138]: Added separate issue https://github.com/ScorexFoundation/sigmastate-interpreter/issues/662 to implement all TODO JITC in a separate PR

File: sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala (hunk 17 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utxo/CostTable.scala]: rolled back all changes in CostTable

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 21 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala]: Rolled back all the changes in reduceToCrypto. New version moved to reduceToCryptoJITC and it is not used by default.

File: sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala [hunk 23]
```diff
@@ -0,0 +1,407 @@
+package sigmastate.utxo
+
+import org.ergoplatform.SigmaConstants
+import sigmastate.{Downcast, Upcast}
+import sigmastate.lang.SigmaParser
+import sigmastate.lang.Terms.OperationId
+
+case class JitCostTable(operCosts: Map[OperationId, Int]) extends (OperationId => Int) {
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala:8]: To reviewer:
This is version is mostly  the same as the old CostTable, but the plan is to make completely new Cost Model which better reflect JIT costs. 
The main invariant which will be guaranteed is that always JIT cost < AOT cost for all non spam scripts.
This shouldn't be hard to do since existing op costs are very conservative and new costs may be reduced proportional to new JIT interpreter speedup. (e.g. 2-3x)

File: sigmastate/src/test/scala/sigmastate/JitCostingSpecification.scala [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+import sigmastate.eval.Sized._
+import sigmastate.eval._
+import sigmastate.helpers.{ContextEnrichingTestProvingInterpreter, ErgoLikeTestInterpreter}
+import sigmastate.interpreter.ContextExtension
+import sigmastate.interpreter.Interpreter.{ScriptNameProp, ScriptEnv, emptyEnv}
+import sigmastate.utxo.{CostTable, JitCostTable}
+import sigmastate.utxo.CostTable._
+import special.sigma.{SigmaTestingData, AvlTree}
+
+class JitCostingSpecification extends SigmaTestingData {
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/JitCostingSpecification.scala:21]: To reviewer: 
This spec will be further updated (in separate PR as part of implementing #665)
The old CostingSpecification remains unchanged since the AOT version should remain unchanged.

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+  val coster: CostAccumulator,
+  val profiler: Profiler,
+  val settings: EvalSettings)
+{
+  /** Evaluates the given expression in the given data environment. */
+  def eval(env: DataEnv, exp: SValue): Any = {
+    exp.evalTo[Any](this, env)
+  }
+
+  def addCostOf(opName: String, opType: SFunc) = {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:69]: please do ScalaDoc for all the public methods

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 26]
```diff
@@ -97,10 +118,21 @@ trait Interpreter extends ScorexLogging {
     (res, currContext.value)
   }
 
+  // TODO after HF: merge with old version (`applyDeserializeContext`)
+  def applyDeserializeContextJITC(context: CTX, exp: Value[SType]): (SigmaPropValue, CTX) = {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:122]: ScalaDoc

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 27]
```diff
@@ -83,6 +97,13 @@ trait Interpreter extends ScorexLogging {
     case x => throw new Error(s"Context-dependent pre-processing should produce tree of type Boolean or SigmaProp but was $x")
   }
 
+  // TODO after HF: merge with old version (`toValidScriptType`)
+  def toValidScriptTypeJITC(exp: SValue): SigmaPropValue = exp match {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:101]: ScalaDoc

File: sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala (hunk 23 above)
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala:8]: ScalaDoc

File: sigmastate/src/main/scala/sigmastate/Values.scala [hunk 28]
```diff
@@ -118,14 +146,89 @@ object Values {
     }
     def notSupportedError(v: SValue, opName: String) =
       throw new IllegalArgumentException(s"Method $opName is not supported for node $v")
+
+    import sigmastate.utxo.JitCostTable
+
+    def costOf(opName: String, opType: SFunc): Int = {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/Values.scala:152]: Please do ScalaDoc for all the public methods

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala [hunk 29]
```diff
@@ -3,9 +3,12 @@ package org.ergoplatform
 import sigmastate.SCollection.SByteArray
 import sigmastate.Values._
 import sigmastate.eval.IRContext
-import sigmastate.interpreter.Interpreter
+import sigmastate.interpreter.{Interpreter, ErgoTreeEvaluator}
+import sigmastate.interpreter.Interpreter.{ScriptEnv, ReductionResult}
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala:7]: fixed

File: sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala [hunk 30]
```diff
@@ -3,9 +3,12 @@ package org.ergoplatform
...
 import sigmastate.utxo._
 
+import scala.util.Try
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/ErgoLikeInterpreter.scala:10]: fixed

File: sigmastate/src/main/scala/sigmastate/eval/BigIntegerOps.scala [hunk 31]
```diff
@@ -26,25 +26,9 @@ object OrderingOps extends LowPriorityOrderingImplicits {
...
-    def toLong(x: BigInteger): Long = x.longValueExact()
-    def toFloat(x: BigInteger): Float = x.floatValue()
-    def toDouble(x: BigInteger): Double = x.doubleValue()
-  }
-  implicit object BigIntegerIsIntegral extends BigIntegerIsIntegral with OrderingOps.BigIntegerOrdering
-
   trait BigIntIsIntegral extends Integral[BigInt] {
     def quot(x: BigInt, y: BigInt): BigInt = x.divide(y)
-    def rem(x: BigInt, y: BigInt): BigInt = x.remainder(y)
+    def rem(x: BigInt, y: BigInt): BigInt = x.mod(y)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/BigIntegerOps.scala]: added comment

File: sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala (hunk 25 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ErgoTreeEvaluator.scala:69]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 26 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:143]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 27 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:106]: this is helper, made it private.

File: sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala (hunk 23 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/utxo/JitCostTable.scala:8]: class removed

File: sigmastate/src/main/scala/sigmastate/Values.scala (hunk 28 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/Values.scala:159]: methods removed

**Comment by aslesarenko** [comment]
The work continues in #699, as the approach to costing will be slightly different. 
All comments are addressed there.
